ocaml/utils/language_extension.ml

```diff
@@ -111,42 +151,95 @@ let is_erasable : type a. a t -> bool = function
   | Module_strengthening ->
       false
 
-module Universe = struct
+module Universe : sig
+  val is_allowed : 'a t -> bool
+  val check : 'a t -> unit
+  val check_maximal : unit -> unit
+
+  type t =
+    | No_extensions
+    | Only_erasable
+    | Any
+
+  val set : t -> bool
+end = struct
   (** Which extensions can be enabled? *)
   type t =
     | No_extensions
     | Only_erasable
     | Any
 
+  let compare t1 t2 =
+    let rank = function
+      | No_extensions -> 1
+      | Only_erasable -> 2
+      | Any -> 3
+    in
+    compare (rank t1) (rank t2)
+
+  let universe = ref Any
+
   let compiler_options = function
     | No_extensions -> "flag -disable-all-extensions"
     | Only_erasable -> "flag -only-erasable-extensions"
     | Any           -> "default options"
 
-  let is_allowed t ext = match t with
+  let is_allowed ext = match !universe with
     | No_extensions -> false
     | Only_erasable -> is_erasable ext
     | Any           -> true
+
+  (* are _all_ extensions allowed? *)
+  let all_allowed () = match !universe with
+    | Any -> true
+    | No_extensions | Only_erasable -> false
+
+  (* Why the [()]? The value restriction. *)
+  let fail () = Format.ksprintf (fun str -> raise (Arg.Bad str))
```

Comment:
I've pushed a commit to refactor this in the style of `Base.failwithf`, which has the advantage that you're less likely to pass more arguments than format specifiers. (Feel free to resolve this.)